azavea/climate-change-lab
# Feature/issue 47 bugfix data requery

## Overview

Refactor the application so that each chart owns its own API requests. This meant mostly gutting the ChartService, which now just has the helper transform methods, and moving the Indicator API HTTP request to a more generic method on IndicatorService. See Notes for additional items of interest changed with this PR.


### Demo

![cclab-chart-owns-data](https://cloud.githubusercontent.com/assets/1818302/18712579/1afde242-7fdd-11e6-8f75-92355fd82aba.png)
![cclab-new-modal](https://cloud.githubusercontent.com/assets/1818302/18712578/1afd2460-7fdd-11e6-8187-647f14070a29.png)

### Notes

- To simplify the climate model selection logic, replaced the checkbox with two buttons that select all/none. Removed the close button and switched to backdrop: static so that the only way to close the modal is to click the update button. Change vetted by @jbranigan 
- To handle re-adding the `backdrop: static` property of the modal, updated to ng2bootstrap 1.1.5 and ng2 2.0.0 release version
- Rather than storing the list of currently selected models in ChartService, `labComponent.project.models` is now the source of truth for the set of currently selected models. The `labComponent.climateModels` property is the list of all available models, and its selected property is kept in sync with the set of models in the `labComponent.project.models` array.
- ChartComponent now fires an event to indicate when one of its settings have changed, so that LabComponent can listen and save the state to localStorage. This fixes the issue where some parts of the UI state would not persist if the page was refreshed.
- The last commit removes a bunch of unused imports throughout the project

## Testing Instructions

 * Stop dev server
 * npm install && npm start
 * Add a few charts from the sidebar, each should independently load data
 * Deleting a chart no longer requeries data for other charts
 * Change anything in the UI, do a full page refresh, state should remain consistent

Note: The `defaultYears` property in `constants.ts` can be safely deleted, it is no longer used.

Closes #36, Closes #47
